### ISSUE TITLE
Web deployment task failed. (The type initializer for 'Microsoft.Web.Deployment.DeploymentManager' threw an exception.)

### ISSUE THREAD

**Issue by dfederm** [dotnet/sdk#11731]
We're seeing this error on some machines but not others. Error:

```
O:\.NetCoreCache\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.Publish\targets\PublishTargets\Microsoft.NET.Sdk.Publish.MSDeployPackage.targets(84,5): Web deployment task failed. (The type initializer for 'Microsoft.Web.Deployment.DeploymentManager' threw an exception.) [X:\bt\1110002\repo\src\Samples\WebSdkPackage\WebSdkPackage.csproj]
```

This task comes from the .NET SDK here:

```
O:\.NetCoreCache\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.Publish\targets\..\tools\net46\Microsoft.NET.Sdk.Publish.Tasks.dll
```

With these dependencies:
```
Package/Publish task Microsoft.NET.Sdk.Publish.Tasks.MsDeploy.VSMSDeploy load assembly Microsoft.Web.Deployment, Version=9.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
Package/Publish task Microsoft.NET.Sdk.Publish.Tasks.MsDeploy.VSMSDeploy load assembly Microsoft.Web.Delegation, Version=7.1.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
```

On working machines, we see these DLLs in the package:
* Microsoft.NET.Sdk.Publish.Tasks.dll
* Microsoft.Web.Delegation.dll
* Microsoft.Web.Deployment.dll
* Microsoft.Web.XmlTransform.dll

On working machines, this DLL is also pulled in (via fusion logs):

```
Microsoft.Web.Deployment.Tracing

LOG: Found assembly by looking in the GAC.
LOG: Binding succeeds. Returns assembly from C:\WINDOWS\Microsoft.Net\assembly\GAC_MSIL\Microsoft.Web.Deployment.Tracing\v4.0_9.0.0.0__31bf3856ad364e35\Microsoft.Web.Deployment.Tracing.dll.
LOG: Assembly is loaded in default load context.
```

So this guy is coming in from the GAC in the successful case.

Since .NET Core is supposedly portable and xplat, I'm not sure why anything would depend on the GAC...

**Comment by joeloff** [comment]
@vijayrkn Is the tracing DLL missing from the package? It seems when Web Deploy is installed from the MSI, everything it needs is in the GAC.

**Comment by dsplaisted** [comment]
> Since .NET Core is supposedly portable and xplat, I'm not sure why anything would depend on the GAC...

This is true when you are using the .NET Core version of the MSBuild, which you get by invoking commands with `dotnet`, ie `dotnet build`.

When you use Full Framework MSBuild which ships with Visual Studio (which happens if you run `msbuild` on the command line or if you open the project in VS), then there are pieces of the portable SDK that are used, but a bunch of stuff comes from Visual Studio (MSBuild, NuGet, the C# compilers, etc.).

So it might be the case that the missing DLL is supposed to be GAC'd by Visual Studio when it installs.  It's also possible that this is a bug in the SDK and the DLL should be included.

**Comment by dfederm** [comment]
@dsplaisted in practice it would be GAC'd for developers, but it also seems like including the dll would be the best practice regardless here.

For example in certain CI environments (eg. CloudBuild), VS is not "installed" but instead just xcopy'ied to the build machine (so there aren't a million separate VS installs on the machine for all versions). This reason may be somewhat historical too as VS used to not even be able to be SxS.

Eliminating dependency on the GAC and registry are worthwhile endeavors regardless though for making the product more robust and portable.

**Comment by davidfowl** [comment]
> Eliminating dependency on the GAC and registry are worthwhile endeavors regardless though for making the product more robust and portable.

I agree with that.

cc @vijayrkn

**Comment by vijayrkn** [comment]
yes, Agreed. 

When WebDeploy was updated to work with mono, WebDeploy team only provided us with these 2 dlls. In NuGet I can see that their official package ships all these 3 dlls - https://www.nuget.org/packages/Microsoft.Web.Deployment/3.6.0. 

Let me check with them to see if the official package contains all these fixes.  If it does, will switch websdk to use the official package which will allow us to redist all these 3 dlls.

**Comment by vijayrkn** [comment]
I have a PR out optimistically - https://github.com/dotnet/websdk/pull/1476. I am waiting on the confirmation from the IIS team. 

Verified the assembly versions in both cases and it is the same so I am not expecting any issues. Looks good from the preliminary tests.
